### PR TITLE
Sector Storage

### PR DESCRIPTION
**Pull request by artyom-yurin**
### Description of the Change
Added Sector Storage
Fixed PoStVerify in VM

### Benefits
Easier work with Sectors

### Possible Drawbacks 
This is a simplified version, it combines several separate classes
It may be necessary to rewrite if the structure becomes more complicated

### CODE REVIEW

File: test/core/sector_storage/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#
+# Copyright Soramitsu Co., Ltd.All Rights Reserved.
+# SPDX - License - Identifier : Apache - 2.0
+#
+
+addtest(sector_storage_test
+        sector_storage_test.cpp)
+
```
Comment (Alexey-N-Chernyshov) [inline test/core/sector_storage/CMakeLists.txt:8]:
```suggestion
        sector_storage_test.cpp
        )
```

File: core/sector_storage/sector_storage_error.hpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  /**
+   * @brief SectorStorage returns these types of errors
+   */
+  enum class SectorStorageError { CANNOT_CREATE_DIR = 1,
+      UNABLE_ACCESS_SEALED_FILE,
+      CANNOT_CLOSE_FILE,
+      CANNOT_REMOVE_DIR,
+      DONOT_MATCH_SIZES,
+      CANNOT_CREATE_FILE,
+      UNKNOWN = 1000 };
```
Comment (Alexey-N-Chernyshov) [inline core/sector_storage/sector_storage_error.hpp:22]:
```suggestion
  enum class SectorStorageError {
    CANNOT_CREATE_DIR = 1,
    UNABLE_ACCESS_SEALED_FILE,
    CANNOT_CLOSE_FILE,
    CANNOT_REMOVE_DIR,
    DONOT_MATCH_SIZES,
    CANNOT_CREATE_FILE,
    UNKNOWN = 1000
  };
```

File: core/primitives/piece/piece_data.hpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+#ifndef CPP_FILECOIN_PIECE_DATA_HPP
+#define CPP_FILECOIN_PIECE_DATA_HPP
+
+#include <string>
+
+namespace fc::primitives::piece {
+
+  class PieceData {
+   public:
+    explicit PieceData(const std::string &path_to_file);
```
Comment (turuslan) [inline core/primitives/piece/piece_data.hpp:15]: Add move constructor to prevent double close.

File: core/primitives/piece/impl/piece_data.cpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+#include "primitives/piece/piece_data.hpp"
+
+#include <fcntl.h>
+#include <boost/filesystem.hpp>
+
+namespace fc::primitives::piece {
+  PieceData::PieceData(const std::string &path_to_file) {
+    fd_ = -1;
```
Comment (turuslan) [inline core/primitives/piece/impl/piece_data.cpp:13]:
```suggestion
  PieceData::PieceData(const std::string &path_to_file) : fd_{-1} {
```

Maybe use named constant for -1, also add bool query if opened or -1

File: core/primitives/sector_file/sector_file.hpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    bool has(const SectorFileType &single_type) const;
+
+   private:
+    int data_;
+  };
+
+  enum SectorFileTypes {
+    FTUnsealed = 1,
+    FTSealed,
+    FTCache,
```
Comment (turuslan) [inline core/primitives/sector_file/sector_file.hpp:35]: This enum is for bitmask
```suggestion
    FTUnsealed = 1,
    FTSealed = 2,
    FTCache = 4,
```

File: core/primitives/sector_file/sector_file.hpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    std::string string() const;
+
+    bool has(const SectorFileType &single_type) const;
+
+   private:
+    int data_;
+  };
+
+  enum SectorFileTypes {
```
Comment (turuslan) [inline core/primitives/sector_file/sector_file.hpp:32]: Should this be `enum class`?

File: core/sector_storage/impl/sector_storage_impl.cpp [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+      size_ = s1.value();
+    }
+  }
+
+  outcome::result<SectorPaths> SectorStorageImpl::acquireSector(
+      SectorId id,
+      const SectorFileType &existing,
+      const SectorFileType &allocate,
+      bool sealing) {
+    auto cache_path =
```
Comment (turuslan) [inline core/sector_storage/impl/sector_storage_impl.cpp:35]: Repetitive. Maybe we can create directory only of requested type, and others will be created lazily when requested?

File: test/core/sector_storage/sector_storage_test.cpp [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+      .sector = 3,
+  };
+
+  EXPECT_OUTCOME_TRUE(
+      paths,
+      sector_storage_->acquireSector(sector, SectorFileTypes::FTCache, 0, true))
+
+  ASSERT_EQ(paths.cache, cache_result);
+  ASSERT_EQ(paths.unsealed, unsealed_result);
+  ASSERT_EQ(paths.sealed, sealed_result);
```
Comment (turuslan) [inline test/core/sector_storage/sector_storage_test.cpp:116]: Wrong test, "cache" is disjoint from "sealed" and "unsealed"

Comment (artyom-yurin) [reply test/core/sector_storage/sector_storage_test.cpp]: Fixed

File: core/sector_storage/impl/sector_storage_impl.cpp (hunk 7 above)
Comment (artyom-yurin) [reply core/sector_storage/impl/sector_storage_impl.cpp]: Fixed

File: core/primitives/sector_file/sector_file.hpp (hunk 5 above)
Comment (artyom-yurin) [reply core/primitives/sector_file/sector_file.hpp]: Fixed

File: core/primitives/piece/impl/piece_data.cpp (hunk 4 above)
Comment (artyom-yurin) [reply core/primitives/piece/impl/piece_data.cpp]: Fixed

File: core/primitives/piece/piece_data.hpp (hunk 3 above)
Comment (artyom-yurin) [reply core/primitives/piece/piece_data.hpp:17]: Added

File: core/sector_storage/sector_storage_error.hpp (hunk 2 above)
Comment (artyom-yurin) [reply core/sector_storage/sector_storage_error.hpp:22]: Fixed

File: test/core/sector_storage/CMakeLists.txt (hunk 1 above)
Comment (artyom-yurin) [reply test/core/sector_storage/CMakeLists.txt:8]: Fixed

File: core/primitives/sector_file/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,9 @@
+# Copyright Soramitsu Co., Ltd. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+add_library(sector_file
+        sector_file.cpp)
+
```
[the remark this answers is not in the document]
Comment (artyom-yurin) [reply core/primitives/sector_file/CMakeLists.txt:6]: Fixed

**Pull request merged**